### PR TITLE
Tighten lint policy and simplify HNSW search loops

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- expand the workspace lint configuration and tighten Clippy thresholds
- refactor HNSW insertion and search helpers to reduce nesting while preserving behaviour
- drop redundant expect allowances in tests and flatten tracing test utilities

## Testing
- make check-fmt
- make lint
- make test

------
https://chatgpt.com/codex/tasks/task_e_68fb4077792483229b1335d6957bb659

## Summary by Sourcery

Tighten lint configuration and refactor core HNSW insertion and search code to reduce complexity, enforce new parameter validations, simplify helper APIs, and bolster test coverage with new edge-case and DataSource tests

New Features:
- Add DataSource batch_distances tests and new HNSW search and insertion edge-case tests

Bug Fixes:
- Reject node attachments above the configured max level in Graph

Enhancements:
- Tighten workspace lint and Clippy thresholds
- Refactor HNSW insertion and search loops to flatten nesting and unify context handling
- Introduce Graph API changes including parameterized with_capacity, attach_node validation, and read_graph/write_graph helpers
- Simplify tracing test utilities to use early returns and remove redundant expect usages
- Adopt 더 builder-style methods for SearchContext and LayerPlanContext

Build:
- Expand workspace.lints in Cargo.toml and enhance clippy.toml with stricter lint rules

Documentation:
- Add module-level documentation to Node storage in hnsw node.rs

Tests:
- Drop redundant expect allowances and update tests to match new Graph API
- Add tests for search respecting minimum ef and DataSource batch distance behavior

### CODE REVIEW

File: chutoro-core/src/hnsw/graph.rs [hunk 1]
```diff
@@ -121,6 +180,14 @@ impl Graph {
     }
 
     pub(crate) fn attach_node(&mut self, node: usize, level: usize) -> Result<(), HnswError> {
+        if level > self.params.max_level() {
+            return Err(HnswError::InvalidParameters {
+                reason: format!(
+                    "level {level} exceeds max_level {}",
+                    self.params.max_level()
+                ),
+            });
```
[the remark this answers is not in the document]
Comment (leynos) [reply chutoro-core/src/hnsw/graph.rs]: Has this now been resolved?

File: chutoro-core/src/hnsw/search.rs [hunk 2]
```diff
@@ -87,165 +77,98 @@ impl Graph {
...
+        best.len() >= ef
+            && best
+                .peek()
+                .is_some_and(|furthest| candidate_distance > furthest.distance)
     }
 
-    #[inline]
-    fn furthest_distance(&self, best: &BinaryHeap<Neighbour>) -> Option<f32> {
-        best.peek().map(|neighbour| neighbour.distance)
+    fn should_enqueue_candidate(best: &BinaryHeap<Neighbour>, ef: usize, distance: f32) -> bool {
```
[the remark this answers is not in the document]
Comment (leynos) [reply chutoro-core/src/hnsw/search.rs]: Has this now been resolved?

File: chutoro-core/src/hnsw/insert.rs [hunk 3]
```diff
@@ -193,28 +177,58 @@ impl Graph {
...
 
         let promote_entry = level > self.entry().map(|entry| entry.level).unwrap_or(0);
+        let max_connections = params.max_connections();
 
         let mut new_node_neighbours = vec![Vec::new(); level + 1];
-        let mut updates: HashMap<(usize, usize), NodeUpdate> = HashMap::new();
+        let mut staged: HashMap<(usize, usize), Vec<usize>> = HashMap::new();
+        let mut needs_trim = HashSet::new();
 
         for layer in plan.layers.into_iter().filter(|layer| layer.level <= level) {
```
[the remark this answers is not in the document]
Comment (leynos) [reply chutoro-core/src/hnsw/insert.rs]: Has this now been resolved?

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[chutoro-core/src/hnsw/tests.rs](https://github.com/leynos/chutoro/pull/32/files/2b42b9b346642177ea37dda76e36b038e36f213c#diff-3b93bd4ba9b354687ec6dee270b2993ffc0b7ab8a8db5861c6477398d8e09ac9)

Comment on lines +440 to +441

~~~rust
            let involves_new_node = left == 2 || right == 2;
            let involves_initial_pair = (left == 0 && right == 1) || (left == 1 && right == 0);
~~~

❌ New issue: [Complex Method](https://codescene.io/projects/71838/delta?repo-id=1205554&review-id=32&biomarker=Complex+Method&filename=chutoro-core%2Fsrc%2Fhnsw%2Ftests.rs&method=non_finite_batch_distance_is_reported)
non_finite_batch_distance_is_reported has a cyclomatic complexity of 11, threshold = 9

**Comment by leynos** [comment]
@coderabbitai pause
